Metrics API

jorgebay opened the pull request:
Expose a metrics interface and create a default implementation that emits those metrics as events.

tolbertam commented on lib/metrics/client-metrics.js:
```diff
@@ -0,0 +1,103 @@
+   * @param {Array<Number>} latency The latency represented in a <code>[seconds, nanoseconds]</code> tuple
+   * Array, where nanoseconds is the remaining part of the real time that can't be represented in second precision.
+   */
+  onSuccessfulResponse(latency) {}
```
I think it would be nice to report also on any response (successful or not) as that would give a better view of how long it takes for the Client to be surfaced a response in general (similar to java-driver's [`Metrics.getRequestsTimer()`](https://docs.datastax.com/en/drivers/java/3.6/com/datastax/driver/core/Metrics.html#getRequestsTimer--).

jorgebay replied:
Good idea!
I'll add it.

tolbertam commented on lib/metrics/default-metrics.js:
```diff
@@ -0,0 +1,175 @@
+ * defaultMetrics.speculativeRetries.on('increment', () => specExecsCount++);
+ * defaultMetrics.successfulResponses.on('increment', latency => specExecsCount++);
+ */
+class DefaultMetrics extends ClientMetrics {
```
This is a neat design because it almost serves as a Dual API.   If a user wants, they can just subscribe on the event emitters, or they can implement their own `ClientMetrics` interface.

tolbertam commented on lib/client-options.js:
```diff
@@ -68,6 +69,7 @@ function defaultOptions () {
     },
     authProvider: null,
     requestTracker: null,
+    metrics: new metrics.DefaultMetrics(),
```
Do we expect much of an added performance cost for emitting events with the `DefaultMetrics` implementation?  I imagine at least initially, users will not use metrics, so if emitting does have a performance cost, maybe we default to a `no-op` implementation where the `onXX` methods don't do anything.

jorgebay replied:
The cost of emitting events is only the function call, [when there are no listeners to the event it can be considered no-op](https://github.com/nodejs/node/blob/v10.13.0/lib/events.js#L176-L179).

tolbertam commented on lib/client-options.js:
```diff
@@ -113,6 +115,10 @@ function extend(baseOptions, userOptions) {
   }
 
+  if (!(options.metrics instanceof metrics.ClientMetrics)) {
+    throw new TypeError('metrics must be an instance of ClientMetrics or null');
```
technically, wouldn't it being null be an issue as we invoke the `onXXX` methods without null checking?  Should it be strictly non-null or a `ClientMetrics` instance?

jorgebay replied:
Good catch, initially I was setting the default instance later but I changed the behaviour and not the error message.
I'll change it.

tolbertam commented on lib/metrics/default-metrics.js:
```diff
@@ -0,0 +1,175 @@
+ * defaultMetrics.errors.on('increment', err => totalErrors++);
+ * defaultMetrics.errors.clientTimeout.on('increment', () => clientTimeoutErrors++);
+ * defaultMetrics.speculativeRetries.on('increment', () => specExecsCount++);
+ * defaultMetrics.successfulResponses.on('increment', latency => specExecsCount++);
```
This should probably be `responseCount++` or maybe `totalLatencySeconds += latency[0]`

jorgebay replied:
🤦

tolbertam commented on test/integration/short/retry-simulator-tests.js:
```diff
@@ -0,0 +1,325 @@
+'use strict';
+
+const assert = require('assert');
```
Excellent test approach!

tolbertam commented on lib/metrics/client-metrics.js:
```diff
@@ -0,0 +1,103 @@
+ * @alias module:metrics~ClientMetrics
+ * @interface
+ */
+class ClientMetrics {
```
I was thinking maybe it would be nice to provide more information as parameters (like Host where error was encountered), but then I think we would start veering into conflating with the Request Tracker implementation and I remembered your l[comment on NODEJS-300](https://datastax-oss.atlassian.net/browse/NODEJS-300?focusedCommentId=50558&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-50558) which covers this.

I think if anyone wants to do something very specific, such as Node-level metrics, they could just use `RequestTracker`, so I like this approach where the API is very simple as it makes it very easy for users to wire in their own functionality.

tolbertam commented on lib/metrics/client-metrics.js:
```diff
@@ -0,0 +1,103 @@
+ * @alias module:metrics~ClientMetrics
+ * @interface
+ */
+class ClientMetrics {
```
Just had a realization.  Currently, all metrics are based on some kind of activity, what about state-based metrics?  For example both the python and java driver expose inflight request counters, total number of connections, connected host count, etc.  Technically those could be event based too, but wonder if sharing come kind of state as a metric is something we should consider.

jorgebay replied:
State info is currently exposed via `client.getState()`. I was also wondering whether we should expose it in the metrics but I didn't see it clearly (should this be refreshed on each call / periodically ?) and the user can always implement a provider specific gauge using `ClientState` information.

tolbertam replied:
> the user can always implement a provider specific gauge using ClientState information.

Just realized that's precisely what I'm doing already :man_facepalming: .  I think I agree, lets keep `ClientMetrics` to only activity-based things, and users can use `Client.getState` to get client & host state.

tolbertam commented:
I wrote my own implementation of `ClientMetrics` and it was super easy and works well!

jorgebay commented:
I think I've addressed all the concerns / suggestions.

Regarding response meters both here and the `RequestTracker`, I found a bug: responses obtained after a speculative execution already "won" are not included or tracked. I've filed [NODEJS-479](https://datastax-oss.atlassian.net/browse/NODEJS-479) for it and I was tempted to try to address it in this PR but I've started digging and it's more complex than it seems. We can tackle it before this release or leave it as a known issue.

tolbertam commented:
Thanks!  I'll give it a final review pass.

```
Regarding response meters both here and the RequestTracker, I found a bug: responses obtained after a speculative execution already "won" are not included or tracked. I've filed NODEJS-479 for it and I was tempted to try to address it in this PR but I've started digging and it's more complex than it seems. We can tackle it before this release or leave it as a known issue.
```

It's a bit subjective, but I think that might be ok.  I believe the java-driver behaves the same way, after a response is received, all subsequent responses are discarded and not recorded, although I do see the point that it's an incomplete view as you may have a slow node and without recording anything your metrics aren't a complete representation.
